src/bidiMapper/modules/emulation/EmulationProcessor.ts

```diff
@@ -0,0 +1,145 @@
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  InvalidArgumentException,
+  UnsupportedOperationException,
+} from '../../../protocol/ErrorResponse';
+import type {
+  EmptyResult,
+  Emulation,
+} from '../../../protocol/generated/webdriver-bidi';
+import type {UserContextStorage} from '../browser/UserContextStorage';
+import type {BrowsingContextImpl} from '../context/BrowsingContextImpl';
+import type {BrowsingContextStorage} from '../context/BrowsingContextStorage';
+
+export class EmulationProcessor {
+  #userContextStorage: UserContextStorage;
+  #browsingContextStorage: BrowsingContextStorage;
+
+  constructor(
+    browsingContextStorage: BrowsingContextStorage,
+    userContextStorage: UserContextStorage,
+  ) {
+    this.#userContextStorage = userContextStorage;
+    this.#browsingContextStorage = browsingContextStorage;
+  }
+
+  async setGeolocationOverride(
+    params: Emulation.SetGeolocationOverrideParameters,
+  ): Promise<EmptyResult> {
+    if ((params.coordinates?.altitude ?? null) !== null) {
+      throw new UnsupportedOperationException(
+        'Geolocation altitude emulation is not supported',
+      );
+    }
+    if ((params.coordinates?.heading ?? null) !== null) {
+      throw new UnsupportedOperationException(
+        'Geolocation heading emulation is not supported',
+      );
+    }
+    if ((params.coordinates?.altitudeAccuracy ?? null) !== null) {
+      throw new UnsupportedOperationException(
+        'Geolocation altitudeAccuracy emulation is not supported',
+      );
+    }
+    if ((params.coordinates?.speed ?? null) !== null) {
+      throw new UnsupportedOperationException(
+        'Geolocation speed emulation is not supported',
+      );
+    }
+
+    const browsingContexts = await this.#getRelatedTopLevelBrowsingContexts(
+      params.contexts,
+      params.userContexts,
+    );
+
+    for (const userContextId of params.userContexts ?? []) {
+      const userContextConfig =
+        this.#userContextStorage.getConfig(userContextId);
+      userContextConfig.emulatedGeolocation = params.coordinates;
+    }
+
+    await Promise.all(
+      browsingContexts.map(
+        async (context) =>
+          await context.cdpTarget.setGeolocationOverride(params.coordinates),
+      ),
+    );
+    return {};
+  }
+
+  /**
+   * Returns a list of top-level browsing contexts.
+   */
+  async #getRelatedTopLevelBrowsingContexts(
```

Comment:
Non blocking:
Can we maybe move this function the UserContextStorage? 
It has the `BrowsingContextStorage` so we can reuse the logic?

Reply:
`UserContextStorage` does not have `BrowsingContextStorage`. It can be extended with one, but it seems to be an odd dependency.